### PR TITLE
Compile PreprocessSpvir.cpp when compiling UnifyIROCL.cpp

### PR DESCRIPTION
**Pull request by troels**
Just a mild annoyance: PreprocessSPVIR.cpp should be compiled whenever UnifyIROCL.cpp is compiled.

### CODE REVIEW

**Comment by pszymich** [comment]
Hello @troels,

Your change does not build with LLVM11 or 12, with following errors reported:

```
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp: In member function 'uint64_t IGC::PreprocessSPVIR::parseSampledImageTy(llvm::StructType*)':
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:45:47: error: conversion from 'llvm::StringRef' to non-scalar type 'std::__cxx11::string {aka std::__cxx11::basic_string<char>}' requested
     std::string Name = sampledImageTy->getName();
                        ~~~~~~~~~~~~~~~~~~~~~~~^~
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp: In member function 'llvm::Value* IGC::PreprocessSPVIR::getWidenImageCoordsArg(llvm::Value*)':
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:82:69: error: 'static llvm::VectorType* llvm::VectorType::get(llvm::Type*, unsigned int)' is deprecated [-Werror=deprecated-declarations]
         Value* undef = UndefValue::get(VectorType::get(coordsType, 4));
                                                                     ^
In file included from ~/src/build_igc/IGC/llvm-deps/src/llvm/include/llvm/IR/Function.h:29:0,
                 from ~/src/build_igc/IGC/llvm-deps/src/llvm/include/llvm/IR/InstVisitor.h:13,
                 from ~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.h:13,
                 from ~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:9:
~/src/build_igc/IGC/llvm-deps/src/llvm/include/llvm/IR/DerivedTypes.h:552:20: note: declared here
 inline VectorType *VectorType::get(Type *ElementType, unsigned NumElements) {
                    ^~~~~~~~~~
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp: In member function 'void IGC::PreprocessSPVIR::visitImageSampleExplicitLod(llvm::CallInst&)':
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:159:92: error: 'static llvm::VectorType* llvm::VectorType::get(llvm::Type*, unsigned int)' is deprecated [-Werror=deprecated-declarations]
     Value* unifiedSampledImage = UndefValue::get(VectorType::get(m_Builder->getInt64Ty(), 3));
                                                                                            ^
In file included from ~/src/build_igc/IGC/llvm-deps/src/llvm/include/llvm/IR/Function.h:29:0,
                 from ~/src/build_igc/IGC/llvm-deps/src/llvm/include/llvm/IR/InstVisitor.h:13,
                 from ~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.h:13,
                 from ~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:9:
~/src/build_igc/IGC/llvm-deps/src/llvm/include/llvm/IR/DerivedTypes.h:552:20: note: declared here
 inline VectorType *VectorType::get(Type *ElementType, unsigned NumElements) {
                    ^~~~~~~~~~
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:210:62: error: 'class llvm::Type' has no member named 'getVectorNumElements'; did you mean 'getStructNumElements'?
             unifiedImageSampleName += std::to_string(dxType->getVectorNumElements());
                                                              ^~~~~~~~~~~~~~~~~~~~
                                                              getStructNumElements
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:220:43: error: 'class llvm::Type' has no member named 'getVectorNumElements'; did you mean 'getStructNumElements'?
     if (retType->isVectorTy() && retType->getVectorNumElements() == 4)
                                           ^~~~~~~~~~~~~~~~~~~~
                                           getStructNumElements
cc1plus: all warnings being treated as errors
make[2]: *** [IGC/CMakeFiles/igc_dll.dir/AdaptorOCL/PreprocessSPVIR.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs...
```

**Comment by troels** [comment]
Hi, @pszymich, yeah it looks like this file has not been moved to LLVM-11.

I've made changes which I've verified works with LLVM-11 and may very well also be enough for LLVM-12.

File: IGC/AdaptorOCL/PreprocessSPVIR.cpp [hunk 1]
```diff
@@ -79,7 +79,7 @@ Value* PreprocessSPVIR::getWidenImageCoordsArg(Value* Coords)
     Value* newCoords = nullptr;
     if (!isa<VectorType>(coordsType))
     {
-        Value* undef = UndefValue::get(VectorType::get(coordsType, 4));
+        Value* undef = UndefValue::get(FixedVectorType::get(coordsType, 4));
```
Comment (pszymich) [inline IGC/AdaptorOCL/PreprocessSPVIR.cpp:82]: Please use `FixedVectorType` wrapper from `IGCLLVM` namespace defined here: https://github.com/intel/intel-graphics-compiler/blob/master/IGC/WrapperLLVM/include/llvmWrapper/IR/DerivedTypes.h#L23

This allows for LLVM <= 10 backwards compatibility.

**Review by pszymich — CHANGES_REQUESTED** [review]

File: IGC/AdaptorOCL/PreprocessSPVIR.cpp (hunk 1 above)
Comment (troels) [reply IGC/AdaptorOCL/PreprocessSPVIR.cpp]: @pszymich, sure. I've changed the calls now.

**Review by pszymich — APPROVED** [review]

**Comment by pszymich** [comment]
Unfortunately still does not build with LLVM12:
```
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:85:59: error: 'unsigned int llvm::VectorType::getNumElements() const' is deprecated: Calling this function via a base VectorType is deprecated. Either call getElementCount() and handle the case where Scalable is true or cast to FixedVectorType. [-Werror=deprecated-declarations]
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:88:79: error: 'unsigned int llvm::VectorType::getNumElements() const' is deprecated: Calling this function via a base VectorType is deprecated. Either call getElementCount() and handle the case where Scalable is true or cast to FixedVectorType. [-Werror=deprecated-declarations]
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:91:72: error: 'unsigned int llvm::VectorType::getNumElements() const' is deprecated: Calling this function via a base VectorType is deprecated. Either call getElementCount() and handle the case where Scalable is true or cast to FixedVectorType. [-Werror=deprecated-declarations]
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:210:73: error: 'unsigned int llvm::VectorType::getNumElements() const' is deprecated: Calling this function via a base VectorType is deprecated. Either call getElementCount() and handle the case where Scalable is true or cast to FixedVectorType. [-Werror=deprecated-declarations]
~/src/igc/IGC/AdaptorOCL/PreprocessSPVIR.cpp:221:33: error: 'unsigned int llvm::VectorType::getNumElements() const' is deprecated: Calling this function via a base VectorType is deprecated. Either call getElementCount() and handle the case where Scalable is true or cast to FixedVectorType. [-Werror=deprecated-declarations]
   221 |    if (VT && VT->getNumElements() == 4)
       |                                 ^
```
For this issue we prefer casting to `IGCLLVM::FixedVectorType`.
Thank you.

**Comment by troels** [comment]
Well, if at first your don't succeed.. :)

I'm unfortunately not able to compile against LLVM-12 from here. But if your last message was the full error list from LLVM-12, I believe it will go through LLVM-12 too, now.

**Comment by pszymich** [comment]
Indeed it did. We've merged your change sucessfully, you can see it here: e2e20fef640b09d2a2c9b4b543dadc13615d2dd4

Thank you for the contribution!

**Pull request closed without merging**